### DISCUSSION TITLE
Routing of our component broke in Joomla 4.

### DISCUSSION THREAD

**Discussion by CeleritasCognito** [joomla/joomla-cms#35621] [General]
I have a very big issue since the installation on my local server of Joomla 4. The routing in our component in Joomla 4 broke. This current router.php creates in Joomla 4 this URL: `localhost/component/eventary/?id=688&Itemid=101`

But it should create this: `localhost/kurse/688-event-name` where "event-name" is the event name. The event name is grabbed by `public function getEventAlias($id)` and automatically updates the URL to the `slug` table inside the database called `eventary_event`. This router.php worked completely fine in Joomla 3.10.0 but broke in Joomla 4.

If I disable the router I get a different URL.

    localhost/component/eventary/?view=event&id=688&format=html&Itemid=101

I also read https://docs.joomla.org/J3.x:Supporting_SEF_URLs_in_your_component but I cannot see what could cause the issue in my current router.

my **router.php**:

```
<?php
/**
 * @package       Eventary
 * @copyright    Copyright (C) 2012 - 2015 Beyounic SA. All rights reserved.
 * @license        GNU GPLv3 <http://www.gnu.org/licenses/gpl.html>
 * @link              http://www.beyounic.com
 */

use Joomla\CMS\Component\Router\RouterBase;
use Joomla\CMS\Application\CMSApplication;
use Joomla\CMS\Factory;

defined('_JEXEC') or die;

/**
 * Eventary Router
 */
class ComEventaryRouter extends JComponentRouterBase
{
    /**
     * Private constructor to avoid direct instantiation
     */
    private function __construct()
    {
    }

    /**
     * Returns an instance of the class
     *
     * @return ComEventaryRouter
     */
    public static function getInstance()
    {
        static $instance;

        if (!$instance) {
            $instance = new ComEventaryRouter();
        }

        return $instance;
    }

    /**
     * Builds a URL from a query object
     *
     * @param array $query query object
     *
     * @return array
     */
    public function build(&$query)
    {
        $segments = [];
        $itemid = null;

        if (empty($query['Itemid'])) {
            return $segments;
        }

        $menu = CMSApplication::getInstance('site')->getMenu()->getItem($query['Itemid']);
        if (!$menu) {
            return $segments;
        }

        $menu_query = $menu->query;

        $menu_view = $menu_query['view'] ?? null;

        if ($menu_view === 'events' || $menu_view === 'calendar') {
            if (array_key_exists('id', $query)) {
                $segments[] = $this->getEventAlias($query['id']);
                unset($query['id'], $query['view'], $query['layout'], $query['tmpl']);
            }
        }

        if (isset($query['view'])) {
            if ($query['view'] === 'checkout' || $query['view'] === 'ticket' || $query['view'] === 'receipt') {
                unset($query['Itemid']);
            } else {
                unset($query['view']);
            }
        }

        unset($query['format']);

        return $segments;
    }

    /**
     * Returns the alias (id-slug) for a document
     *
     * @param string $slug document slug
     *
     * @return mixed
     */
    public function getEventAlias($id)
    {
        $db = Factory::getDbo();
        $db->setQuery(sprintf("
            SELECT CONCAT_WS('-', tbl.eventary_event_id, tbl.slug) AS alias
            FROM `#__eventary_events` as tbl
            WHERE tbl.eventary_event_id = %s", $db->quote($id)));

        return $db->loadResult();
    }

    /**
     * Parse the segments into query string
     *
     * @param array $segments
     *
     * @return array
     */
    public function parse(&$segments)
    {
        // Circumvent Joomla's auto encoding
        foreach ($segments as &$segment) {
            $segment = urldecode($segment);
            $pos = strpos($segment, ':');
            if ($pos !== false) {
                $segment[$pos] = '-';
            }

        }

        unset($segment); // break the reference with the last element

        $vars = [];
        $app = Factory::getApplication();

        $menu = $app->getMenu()->getActive();
        $view = $app->input->getCmd('view', null);
        $event = $this->_findEvent(array_slice(array_reverse($segments), 0, 2));

        if ($event) {
            $position = array_search($event, $segments);
            $view = 'event';

            unset($segments[$position]);
        }

        if ($menu !== null) {
            if ($view === 'event') {
                $vars['view'] = $view;

                if (isset($menu->query['view']) && $menu->query['view'] === 'event') {
                    $vars['slug'] = $menu->query['slug'];
                } elseif ($event) {
                    $pieces = explode('-', $event, 2);
                    $vars['slug'] = array_pop($pieces);

                } else {
                    $vars['slug'] = '';
                }

                $vars['path'] = '';
                if (isset($menu->query['path'])) {
                    $vars['path'] .= $menu->query['path'] . '/';
                }

                $vars['path'] .= implode('/', $segments);
            } else // list view
            {
                $vars['view'] = $menu->query['view'];
                $vars['slug'] = array_pop($segments);
                if (strpos('/', $vars['slug']) !== false) {
                    $pieces = explode('/', $vars['slug']);
                    $vars['slug'] = array_pop($pieces);
                }

                if (isset($menu->query['layout'])) {
                    $vars['layout'] = $menu->query['layout'];
                }
            }
        }


        return $vars;
    }

    /**
     * Checks the last two segments and tries to find an event
     *
     * @param array $segments
     * @return null|string
     */
    protected function _findEvent($segments)
    {
        $result = null;

        foreach ($segments as $segment) {
            if (preg_match('#[0-9]+-(.*?)#', $segment) && $this->_isEventAlias($segment)) {
                $result = $segment;
            }
        }
        return $result;
    }

    /**
     * Returns true if the given string is an event alias
     *
     * @param string $alias event alias
     *
     * @return bool
     */
    protected function _isEventAlias($alias)
    {
        [$id, $slug] = explode('-', $alias, 2);
        $query = 'SELECT COUNT(*) FROM #__eventary_events'
            . ' WHERE eventary_event_id = %d AND slug = %s';

        $db = Factory::getDbo();
        $db->setQuery(sprintf($query, $id, $db->quote($slug)));

        return (bool)$db->loadResult();
    }
}

/**
 * Hooks up Eventary router to Joomla URL build event
 *
 * @param array $query
 *
 * @return array
 */
function EventaryBuildRoute(&$query)
{
    return ComEventaryRouter::getInstance()->build($query);

}

/**
 * Hooks up Eventary router to Joomla URL parse event
 *
 * @param array $segments
 *
 * @return array
 */
function EventaryParseRoute($segments)
{
    return ComEventaryRouter::getInstance()->parse($segments);

}
```

**Comment by joomdonation** [comment]
I can only guess that it happens because in Joomla 3, if you do not pass Itemid in the URL of Router::_ call, Joomla will use pass current active menu item to the URI object for building SEF url. 

However, in Joomla 4, for some reasons, that logic is not used anymore. The method `createUri ` https://github.com/joomla/joomla-cms/blob/3.10-dev/libraries/src/Router/SiteRouter.php#L709 from `SiteRouter` class was removed and it causes the issue. Maybe you can add that method back to SiteRouter class in your Joomla 4 installation to see if it solves your issue?

**Comment by Fedik** [comment]
You have to find correct Itemid for the link in your router,
In joomla core it handled by [`Joomla\CMS\Component\Router\Rules\MenuRules` ](https://github.com/joomla/joomla-cms/blob/f575a49103e29e94f1d62dfa23b4ae53e680879b/libraries/src/Component/Router/Rules/MenuRules.php#L22) 
while `MenuRules::preprocess()` stage.

Maybe can try look in com_content as example of new router.
https://github.com/joomla/joomla-cms/blob/f575a49103e29e94f1d62dfa23b4ae53e680879b/components/com_content/src/Service/Router.php#L100-L102

**Reply by Fedik** [reply]
Upd: add preprocess() to your router, and look for correct Itemid there

**Reply by joomdonation** [reply]
@Fedik The problem here is we made some changes to router process in Joomla 4. The logic to dynamic set value for option and Itemid base on current active URL was removed https://github.com/joomla/joomla-cms/blob/3.10-dev/libraries/src/Router/SiteRouter.php#L709. 

Maybe there was reasons for that changes but it makes url for third party extensions broken when upgrading to Joomla 4.

**Reply by Fedik** [reply]
I think it was removed because now it should be `MenuRules` which is searching for correct Item id, based on active URL ;) 

Hm, and that old code does not looks very logical, I wonder if it even worked :)

**Reply by joomdonation** [reply]
The problem is not all extensions use the new routing code. That old code works (I even see tests for these cases when I debugged something related to routing  few years ago :) ).

**Reply by Fedik** [reply]
That is true, but not sure what can be done here.
I would just try to create `preprocess()` thing in component router, and try to lookup menu for correct Itemid.

**Reply by joomdonation** [reply]
I think the easiest way for this case would be getting Itemid from request, check if it is a menu item from this component, append &Itemid=CURRENT_ITEM_ID to the link in **Route::_** call. That would make it behaves in the same way with J3 without (if I understand his problem correctly)

**Reply by CeleritasCognito** [reply]
@joomdonation Well, the issue I have is that without the router. (renamed to _router.php) I get that URL:
`localhost/component/eventary/?view=event&id=688&format=html&Itemid=101`
with it enabled I get `localhost/component/eventary/?id=688&Itemid=101` but only the disabled router.php redirects correctly to the views. But the URLs should look like that `localhost/kurse/688-event-name`. I also added the `createUri` back but that didn't fix the issue.

**Reply by joomdonation** [reply]
@YoKoGFX Without seeing the full source code of the extension, we can only guess :). Somehow, you will have to find a way to find the right Itemid for the link. For my extensions, I still use a helper router class (the old way) and it still works well for me.

**Reply by Fedik** [reply]
@YoKoGFX in your ComEventaryRouter add preprocess:

```php
public function preprocess(&$query)
{
  $menu = CMSApplication::getInstance('site')->getMenu();

  $searchLink = 'index.php?option=com_eventary&view=event&id=68'; // based on $query
  $candidates = $menu->getItems('link', $searchLink); // Search for a menu that suits your link

  if (!$candidates) return; // Nothing found

  // ... let say first candidat is suits you:
 $query['Itemid'] = $candidates[0]->id;

}
```

More general:
```php
public function preprocess(&$query)
{
  $menu = CMSApplication::getInstance('site')->getMenu();

  // Search for all menu items for your component
  $candidates = $menu->getItems('component', 'com_eventary'); 

  if (!$candidates) return; // Nothing found

  // Check each if it suits current $query
  foreach ($candidates as $candidate) 
  {
    if ( /* .. $candidate is goood */) 
    {
      $query['Itemid'] = $candidate->id;
      break;
    }
  }
}
```

This is very basic, but should be good enough if you do not want to rewrite your router to use RouterRules.
